src/monodroid/jni/monodroid-glue.c

```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
 	return 1;
 }
 
+// Add a reference from an IGCUserPeer jobject to another jobject
 static mono_bool
-add_reference (JNIEnv *env, MonoObject *obj, MonoJavaGCBridgeInfo *bridge_info, MonoObject *reffed_obj)
+add_reference_jobject (JNIEnv *env, jobject handle, jobject reffed_handle)
 {
 	jclass java_class;
 	jmethodID add_method_id;
-	void *handle, *reffed_handle;
-#if DEBUG
-	MonoClass *klass;
-	klass = mono.mono_object_get_class (obj);
-#endif
 
-	mono.mono_field_get_value (obj, bridge_info->handle, &handle);
 	java_class = (*env)->GetObjectClass (env, handle);
 	add_method_id = (*env)->GetMethodID (env, java_class, "monodroidAddReference", "(Ljava/lang/Object;)V");
 	if (add_method_id) {
-		mono.mono_field_get_value (reffed_obj, bridge_info->handle, &reffed_handle);
```

Comment:
Notice there is a pretty big bug here which this patch fixes. The same bridge_info is used for obj and reffed_obj. If obj is an Object and reffed_obj is a Throwable, this could lead to Something Very Bad happening.